hack/e2e.go

```diff
@@ -14,968 +14,138 @@ See the License for the specific language governing permissions and
-			report = append(report, matches[1])
-		}
-		if strings.HasPrefix(l, "+") && len(strings.TrimPrefix(l, "+")) > 0 {
-			added = append(added, l)
-			report = append(report, l)
-		}
-	}
-	if len(added) > 0 {
-		return fmt.Errorf("Error: %d leaked resources\n%v", len(added), strings.Join(report, "\n"))
-	}
-	return nil
-}
-
-func ListResources() ([]byte, error) {
-	log.Printf("Listing resources...")
-	cmd := exec.Command("./cluster/gce/list-resources.sh")
-	if *verbose {
-		cmd.Stderr = os.Stderr
-	}
-	stdout, err := cmd.Output()
-	if err != nil {
-		return nil, fmt.Errorf("Failed to list resources (%s):\n%s", err, string(stdout))
-	}
-	return stdout, nil
-}
-
-func Build() error {
-	// The build-release script needs stdin to ask the user whether
-	// it's OK to download the docker image.
-	cmd := exec.Command("make", "quick-release")
-	cmd.Stdin = os.Stdin
-	if err := finishRunning(cmd); err != nil {
-		return fmt.Errorf("error building kubernetes: %v", err)
-	}
-	return nil
-}
-
-type deployer interface {
-	Up() error
-	IsUp() error
-	SetupKubecfg() error
-	Down() error
-}
-
-func getDeployer() (deployer, error) {
-	switch *deployment {
-	case "bash":
-		return bash{}, nil
-	case "kops":
-		return NewKops()
-	case "kubernetes-anywhere":
-		return NewKubernetesAnywhere()
-	default:
-		return nil, fmt.Errorf("Unknown deployment strategy %q", *deployment)
-	}
+	return c.Wait()
 }
 
-type bash struct{}
-
-func (b bash) Up() error {
-	return finishRunning(exec.Command("./hack/e2e-internal/e2e-up.sh"))
+// Struct that allows unit tests to override functionality.
+type tester struct {
+	// os.Stat
+	stat func(string) (os.FileInfo, error)
+	// exec.LookPath
+	lookPath func(string) (string, error)
+	// build.Default.GOPATH
+	goPath string
+	wait   func(string, ...string) error
 }
 
-func (b bash) IsUp() error {
-	return finishRunning(exec.Command("./hack/e2e-internal/e2e-status.sh"))
+func newTester() tester {
+	return tester{os.Stat, exec.LookPath, build.Default.GOPATH, wait}
 }
 
-func (b bash) SetupKubecfg() error {
-	return nil
-}
-
-func (b bash) Down() error {
-	return finishRunning(exec.Command("./hack/e2e-internal/e2e-down.sh"))
-}
-
-type kops struct {
-	path        string
-	kubeVersion string
-	sshKey      string
-	zones       []string
-	nodes       int
-	adminAccess string
-	cluster     string
-	kubecfg     string
-}
-
-func NewKops() (*kops, error) {
-	if *kopsPath == "" {
-		return nil, fmt.Errorf("--kops must be set to a valid binary path for kops deployment.")
-	}
-	if *kopsCluster == "" {
-		return nil, fmt.Errorf("--kops-cluster must be set to a valid cluster name for kops deployment.")
-	}
-	if *kopsState == "" {
-		return nil, fmt.Errorf("--kops-state must be set to a valid S3 path for kops deployment.")
-	}
-	sshKey := *kopsSSHKey
-	if sshKey == "" {
-		usr, err := user.Current()
-		if err != nil {
-			return nil, err
+// Try to find kubetest, either GOPATH/bin/kubetest or PATH
+func (t tester) look() (string, error) {
```

Comment:
`look` isn't a very descriptive name... I'd suggest `findKubetest` or something like that.